File: ruby/sum-of-multiples/sum_of_multiples.rb
class SumOfMultiples
  LIMIT_TO_RANGE = ->(limit) { (1...limit) }
  private_constant :LIMIT_TO_RANGE

  def initialize(*numbers)
    @numbers = numbers.freeze
  end

  def to(limit)
    limit
      .then(&LIMIT_TO_RANGE)
      .select(&method(:multiples?))
      .sum
  end

  private

  attr_reader :numbers

  def multiples?(range_number)
    numbers.any? { |number| multiple?(range_number, number) }
  end

  def multiple?(number1, number2)
    (number1 % number2).zero?
  end
end

File: elixir/sum-of-multiples/lib/sum_of_multiples.ex
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    1..(limit - 1)
    |> Stream.filter(&multiples?(factors, &1))
    |> Enum.sum()
  end

  defp multiples?(factors, range_number) do
    Enum.any?(factors, &multiple?(range_number, &1))
  end

  defp multiple?(range_number, factor), do: rem(range_number, factor) == 0
end
